File: xim/resource/table/ZoneDecrypt.kt
package xim.resource.table

import js.typedarrays.Uint8Array
import xim.resource.ByteReader
import xim.resource.SectionHeader
import xim.util.OnceLogger

object ZoneDecrypt {

    private val keyTable1: Uint8Array by lazy { MainDll.getZoneDecryptTable1() }
    private val keyTable2: Uint8Array by lazy { MainDll.getZoneDecryptTable2() }

    fun decryptZoneObjects(sectionHeader: SectionHeader, byteReader: ByteReader) {
        byteReader.offsetFrom(sectionHeader, 0x10)

        val metadata = byteReader.next32()
        var decodeLength = metadata and 0x00FFFFFF
        val mode = (metadata shr 24) and 0xFF

        if (mode <= 0x1A) {
            return
        }

        val keyNodeData = byteReader.next32()
        val nodeCount = keyNodeData and 0x00FFFFFF

        val keyIndex = (keyNodeData shr 24) and 0xFF

        var key = keyTable1[keyIndex xor 0xFF].toInt()
        var keyCounter  = 0
        var consumed = 0

        if (byteReader.position + decodeLength > sectionHeader.nextSectionOffset()) {
            val extra = (byteReader.position + decodeLength) - sectionHeader.nextSectionOffset()
            OnceLogger.warn("[${sectionHeader.sectionId}] Decode length was too long by: 0x${extra.toString(0x10)}")
            decodeLength -= extra
        }

        while (consumed < decodeLength) {
            val xorLength = ((key shr 4) and 7) + 16

            val shouldApplyMask = (key and 1) != 0
            val hasRemainingLength = consumed + xorLength < decodeLength

            if (shouldApplyMask && hasRemainingLength) {
                for (i in 0 until xorLength) {
                    byteReader.xorNext(0xFF.toByte())
                }
            } else {
                byteReader.position += xorLength
            }

            keyCounter += 1
            key += keyCounter
            consumed += xorLength
        }

        // Names are masked too
        byteReader.offsetFrom(sectionHeader, 0x30)

        val pos = byteReader.position
        for (i in 0 until nodeCount) {
            byteReader.position = pos + i * 0x64
            for (j in 0 until 0x10) { byteReader.xorNext(0x55) }
        }
    }

    fun decryptZoneMesh(sectionHeader: SectionHeader, byteReader: ByteReader, reencrypt: Boolean = false) {
        if (reencrypt) {
            decryptZoneMeshSecondPass(sectionHeader, byteReader)
            decryptZoneMeshFirstPass(sectionHeader, byteReader)
        } else {
            decryptZoneMeshFirstPass(sectionHeader, byteReader)
            decryptZoneMeshSecondPass(sectionHeader, byteReader)
        }

        byteReader.offsetFrom(sectionHeader, 0x10)
    }

    private fun decryptZoneMeshFirstPass(sectionHeader: SectionHeader, byteReader: ByteReader) {
        byteReader.offsetFrom(sectionHeader, 0x10)

        // 0123
        val metadata = byteReader.next32()
        val totalSize = metadata and 0x00FFFFFF
        val decodeLength = totalSize - 0x8          // 8 bytes of header data, rest is decoded

        val mode = (metadata shr 24) and 0xFF
        if (mode < 0x5) {
            return
        }

        // 4
        byteReader.next8() // not decrypt related

        // 5
        val keyIndex = byteReader.next8()

        // 67 - unused in first pass
        byteReader.next16()

        var key = keyTable1[keyIndex xor 0xF0].toInt()
        var keyCounter = 0

        for (i in 0 until decodeLength) {
            val keyMod = ((key and 0xFF) shl 8) or (key and 0xFF)
            key += ++keyCounter

            val mask = (keyMod ushr (key and 0x7)).toByte()
            byteReader.xorNext(mask)
            key += ++keyCounter
        }
    }

    private fun decryptZoneMeshSecondPass(sectionHeader: SectionHeader, byteReader: ByteReader) {
        byteReader.offsetFrom(sectionHeader, 0x10)

        // 0123
        val metadata = byteReader.next32()
        val totalSize = metadata and 0x00FFFFFF
        val decodeLength = totalSize - 0x8          // 8 bytes of header data, rest is decoded

        // 4
        byteReader.next8() // not decrypt related

        // 5
        var key1 = byteReader.next8() xor 0xF0
        var key2 = keyTable2[key1].toInt()

        val decodeCount = (decodeLength and (0xF.inv())) / 2

        // 67
        val needsSecondPass = byteReader.next16() == 0xFFFF
        if (!needsSecondPass) {
            return
        }

        // Swap
        for (i in 0 until decodeCount step 8) {
            if ((key2 and 1) == 1) {
                byteReader.swapNext8(decodeCount, 8)
            } else {
                byteReader.position += 0x8
            }

            key1 += 9
            key2 += key1
        }
    }

}

File: lib/resource/table/zoneDecrypt.ts
/**
 * Zone mesh and zone object decryption.
 *
 * Ported from xim/resource/table/ZoneDecrypt.kt.
 * Uses two 256-byte key tables extracted from FFXiMain.dll.
 */

import type { ByteReader, SectionOffsetSource } from '../byteReader'
import type { MainDll } from './mainDll'

let keyTable1: Uint8Array | null = null
let keyTable2: Uint8Array | null = null

export function initZoneDecrypt(mainDll: MainDll): void {
  keyTable1 = mainDll.getZoneDecryptTable1()
  keyTable2 = mainDll.getZoneDecryptTable2()
}

function requireKeyTable1(): Uint8Array {
  if (!keyTable1) throw new Error('ZoneDecrypt not initialized. Call initZoneDecrypt() first.')
  return keyTable1
}

function requireKeyTable2(): Uint8Array {
  if (!keyTable2) throw new Error('ZoneDecrypt not initialized. Call initZoneDecrypt() first.')
  return keyTable2
}

/**
 * Decrypt zone object data (section 0x1C).
 * XOR-based decryption with variable-length chunks, then unmask object names.
 */
export function decryptZoneObjects(sectionHeader: SectionOffsetSource, byteReader: ByteReader): void {
  const kt1 = requireKeyTable1()

  byteReader.offsetFrom(sectionHeader, 0x10)

  const metadata = byteReader.next32()
  let decodeLength = metadata & 0x00ffffff
  const mode = (metadata >>> 24) & 0xff

  if (mode <= 0x1a) {
    return
  }

  const keyNodeData = byteReader.next32()
  const nodeCount = keyNodeData & 0x00ffffff
  const keyIndex = (keyNodeData >>> 24) & 0xff

  let key = kt1[keyIndex ^ 0xff]!
  let keyCounter = 0
  let consumed = 0

  const sectionEnd = sectionHeader.sectionStartPosition + sectionHeader.sectionSize

  if (byteReader.position + decodeLength > sectionEnd) {
    const extra = (byteReader.position + decodeLength) - sectionEnd
    console.warn(`[${sectionHeader}] Decode length too long by: 0x${extra.toString(16)}`)
    decodeLength -= extra
  }

  while (consumed < decodeLength) {
    const xorLength = ((key >>> 4) & 7) + 16

    const shouldApplyMask = (key & 1) !== 0
    const hasRemainingLength = consumed + xorLength < decodeLength

    if (shouldApplyMask && hasRemainingLength) {
      for (let i = 0; i < xorLength; i++) {
        byteReader.xorNext(0xff)
      }
    } else {
      byteReader.position += xorLength
    }

    keyCounter += 1
    // Kotlin uses unbounded 32-bit int addition -- no masking
    key += keyCounter
    consumed += xorLength
  }

  // Names are masked too
  byteReader.offsetFrom(sectionHeader, 0x30)
  const pos = byteReader.position

  for (let i = 0; i < nodeCount; i++) {
    byteReader.position = pos + i * 0x64
    for (let j = 0; j < 0x10; j++) {
      byteReader.xorNext(0x55)
    }
  }
}

/**
 * Decrypt zone mesh data (section 0x2E).
 * Two-pass: XOR cipher + conditional byte-swap.
 */
export function decryptZoneMesh(sectionHeader: SectionOffsetSource, byteReader: ByteReader): void {
  decryptZoneMeshFirstPass(sectionHeader, byteReader)
  decryptZoneMeshSecondPass(sectionHeader, byteReader)
  byteReader.offsetFrom(sectionHeader, 0x10)
}

function decryptZoneMeshFirstPass(sectionHeader: SectionOffsetSource, byteReader: ByteReader): void {
  const kt1 = requireKeyTable1()

  byteReader.offsetFrom(sectionHeader, 0x10)

  const metadata = byteReader.next32()
  const totalSize = metadata & 0x00ffffff
  const decodeLength = totalSize - 0x8

  const mode = (metadata >>> 24) & 0xff
  if (mode < 0x5) {
    return
  }

  byteReader.next8() // not decrypt related
  const keyIndex = byteReader.next8()
  byteReader.next16() // unused in first pass

  // Kotlin uses unbounded 32-bit int -- no masking on key progression.
  // JavaScript numbers are 64-bit floats but safe for integers up to 2^53.
  let key = kt1[keyIndex ^ 0xf0]!
  let keyCounter = 0

  for (let i = 0; i < decodeLength; i++) {
    const keyMod = ((key & 0xff) << 8) | (key & 0xff)
    key += ++keyCounter

    const mask = (keyMod >>> (key & 0x7)) & 0xff
    byteReader.xorNext(mask)
    key += ++keyCounter
  }
}

function decryptZoneMeshSecondPass(sectionHeader: SectionOffsetSource, byteReader: ByteReader): void {
  const kt2 = requireKeyTable2()

  byteReader.offsetFrom(sectionHeader, 0x10)

  const metadata = byteReader.next32()
  const totalSize = metadata & 0x00ffffff
  const decodeLength = totalSize - 0x8

  byteReader.next8() // not decrypt related
  let key1 = byteReader.next8() ^ 0xf0
  let key2 = kt2[key1 & 0xff]!

  const decodeCount = (decodeLength & ~0xf) / 2

  const needsSecondPass = byteReader.next16() === 0xffff
  if (!needsSecondPass) {
    return
  }

  // Swap
  for (let i = 0; i < decodeCount; i += 8) {
    if ((key2 & 1) === 1) {
      byteReader.swapNext8(decodeCount, 8)
    } else {
      byteReader.position += 0x8
    }

    // Kotlin uses unbounded 32-bit int -- no masking
    key1 += 9
    key2 += key1
  }
}
